File: 05-Algorithms/code/typescript/src/dfs.ts
import { buildAdjList, checkVertexInRange, type Edge } from "./graphCommon.js";

/**
 * Depth-first traversal from `src` over n vertices (0..n-1). Returns
 * vertices in preorder (a vertex is appended the moment it is first
 * visited, before recursing into its neighbors). Neighbors are always
 * explored in ascending vertex-id order, so the result is fully
 * deterministic. Implemented iteratively with an explicit stack to avoid
 * recursion-depth limits. O(V + E).
 */
export function dfs(n: number, edges: Edge[], src: number, directed = false): number[] {
  checkVertexInRange(n, src);
  const adj = buildAdjList(n, edges, directed);
  const visited = new Array(n).fill(false);
  const order: number[] = [];
  const stack: number[] = [src];
  while (stack.length > 0) {
    const u = stack.pop() as number;
    if (visited[u]) continue;
    visited[u] = true;
    order.push(u);
    const neighbors = adj[u];
    for (let i = neighbors.length - 1; i >= 0; i--) {
      if (!visited[neighbors[i]]) stack.push(neighbors[i]);
    }
  }
  return order;
}

File: 05-Algorithms/code/python/seb_alg/dfs.py
"""Mirrors `typescript/src/dfs.ts` exactly."""

from __future__ import annotations

from typing import List, Sequence

from .graph_common import build_adj_list, check_vertex_in_range


def dfs(n: int, edges: Sequence[Sequence[int]], src: int, directed: bool = False) -> List[int]:
    """Depth-first traversal from src, preorder. Iterative with explicit
    stack to avoid recursion-depth limits. O(V + E).
    """
    check_vertex_in_range(n, src)
    adj = build_adj_list(n, edges, directed)
    visited = [False] * n
    order: List[int] = []
    stack: List[int] = [src]
    while stack:
        u = stack.pop()
        if visited[u]:
            continue
        visited[u] = True
        order.append(u)
        neighbors = adj[u]
        for i in range(len(neighbors) - 1, -1, -1):
            if not visited[neighbors[i]]:
                stack.append(neighbors[i])
    return order
